### PR TITLE
GEOPY-614

### PR DESCRIPTION
**Pull request by benk-mira**
**GEOPY-614 - Create utility to index survey objects by line**

### CODE REVIEW

**Review by domfournier — CHANGES_REQUESTED** [review]
I think we should revise the also as it will be pretty slow and prone to fail if you have overlapping lines. We can talk about it.

File: geoapps/utils/surveys.py [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+
+
+def find_unique_tops(locations):
+
+    locs, inds = np.unique(locations[:, :2], axis=0, return_inverse=True)
+
+    if len(locs) != len(locations):
+        tops = []
+        for i in range(len(locs)):
+            tops.append(locations[:, 2][np.where(inds == i)[0]].max())
```
[the remark this answers is not in the document]
Comment (benk-mira) [reply geoapps/utils/surveys.py:49]: I generalized the function to handle 2D sections so I could re-use elsewhere

File: geoapps/utils/surveys.py [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+
+import numpy as np
+from geoh5py.data import FloatData
+from geoh5py.objects import CurrentElectrode, PotentialElectrode
+from scipy.spatial import cKDTree
+
+from geoapps.utils.statistics import is_outlier
+
+
+def new_neighbors(distances, neighbors, nodes):
```
[the remark this answers is not in the document]
Comment (benk-mira) [reply geoapps/utils/surveys.py:19]: done

File: geoapps/utils/surveys.py [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+    """index into neighbor arrays excluding zero distance and past neighbors."""
```
[the remark this answers is not in the document]
Comment (benk-mira) [reply geoapps/utils/surveys.py:20]: done

[the remark this answers is not in the document]
Comment (benk-mira) [reply geoapps/utils/surveys.py:20]: Done, sorry

File: geoapps/utils/surveys.py [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+def survey_lines(survey, start_loc, save: str | None = None):
+    """Build an array of line ids for a survey laid out in a line biased grid."""
+
+    # extract xy locations and create linear indexing
+    locs = survey.vertices[:, :2]
+    nodes = np.arange(len(locs)).tolist()
+
+    # find the id of the closest point to the starting location
+    start_id = np.argmin(
+        np.sqrt(((locs[:, 0] - start_loc[0]) ** 2) + ((locs[:, 1]) - start_loc[1]) ** 2)
```
[the remark this answers is not in the document]
Comment (benk-mira) [reply geoapps/utils/surveys.py:89]: Indeed, I forgot about the axis option.  Thanks!

File: geoapps/utils/surveys.py [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+        neighbors = neighbors[new_ids]
+        next_id = np.argmin(distances)
+        return distances[next_id], neighbors[next_id]
+
+    else:
+        return next_neighbor(tree, point, nodes, n + 3)
+
+
+def find_unique_tops(locations):
+
```
[the remark this answers is not in the document]
Comment (benk-mira) [reply geoapps/utils/surveys.py:44]: done

File: geoapps/utils/surveys.py [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+    ind = [
+        i in nodes if distances[neighbors.tolist().index(i)] != 0 else False
+        for i in neighbors
+    ]
+    return np.where(ind)[0].tolist()
+
+
+def next_neighbor(tree, point, nodes, n=3):
+    """Returns smallest distance neighbor that has not yet been traversed"""
+    distances, neighbors = tree.query(point, n)
```
[the remark this answers is not in the document]
Comment (domfournier) [reply geoapps/utils/surveys.py:30]: Correct. New issue to work on this later GEOPY-638

**Review by domfournier — APPROVED** [review]
Grand

**Review by domfournier — APPROVED** [review]

**Pull request merged**